src/vs/base/node/userSettings.ts

```diff
@@ -90,8 +90,35 @@ export class UserSettings {
 	}
 
 	private registerWatchers(): void {
-		this.watcher = fs.watch(path.dirname(this.appSettingsPath));
-		this.watcher.on('change', (eventType: string, fileName: string) => this.onSettingsFileChange(eventType, fileName));
+		let self = this;
+		function attachSettingsChangeWatcher(watchPath: string): void {
+			self.watcher = fs.watch(watchPath);
+			self.watcher.on('change', (eventType: string, fileName: string) => self.onSettingsFileChange(eventType, fileName));
+		}
+
+		// Attach a watcher to the settings directory
+		attachSettingsChangeWatcher(path.dirname(this.appSettingsPath));
+
+		// Follow symlinks for settings and keybindings and attach watchers if they resolve
+		let followSymlinkPaths = [
+			this.appSettingsPath,
+			this.appKeybindingsPath
+		];
+		followSymlinkPaths.forEach((path) => {
+			fs.lstat(path, function(err, stats) {
+				if (err) {
+					return;
+				}
+				if (stats.isSymbolicLink()) {
+					fs.readlink(path, function(err, realPath) {
```

Comment:
What if the symlink is actually a folder? Can you check that the stat is a file?